File: arrays_hashing/encode_and_decode_strings/encode_decode_strings.go
// Package encodedecode will solve the encode and decode strings problem.
package encodedecode

import (
	"strconv"
	"strings"
)

type Solution struct{}

func (s *Solution) Encode(strs []string) string {
	result := ""
	for _, str := range strs {
		// Take the word and concat "length#word" as the encoding.
		result += strconv.Itoa(len(str)) + "#" + str
	}
	return result
}

func (s *Solution) Decode(encoded string) []string {
	if len(encoded) == 0 {
		return []string{}
	}
	var res []string
	for i := 0; i < len(encoded); {
		// Find the hash tag.
		pos := strings.Index(encoded[i:], "#") + i
		// Use i position and the hashtag index to find the length of the word.
		length, _ := strconv.Atoi(encoded[i:pos])
		// Now using slicing to grab the word.
		res = append(res, encoded[pos+1:pos+1+length])
		// Make sure the loop starts at the end of the word.
		i = pos + 1 + length
	}
	return res
}

File: arrays_hashing/encode_and_decode_strings/encode_decode_strings.py
def encode(strs):
    result = ""
    for word in strs:
        result += str(len(word)) + "#" + word
    return result


def decode(str):
    result = []
    i = 0

    while i < len(str):
        # Start j where i is at.
        j = i
        while str[j] != "#":
            # Find the end of the word.
            j += 1
        # i is the start of the number and j is the index of the hashtag.
        length = int(str[i:j])
        # j + 1 would be the char after the hashtag.
        i = j + 1
        # now the i is the first char plus the length is the end of the world.
        j = i + length
        # Now slice from i:j to grab the word.
        result.append(str[i:j])
        # Start the i at the end of word so you can capture the next word.
        i = j
    return result
